### DISCUSSION TITLE
Tagged pdfs <extract data using StructTreeRoot>

### DISCUSSION THREAD

**Discussion by mllife** [pikepdf/pikepdf#614] [Q&A]
I am working with pdfs using pymupdf for some time, but recently came across tagged pdfs and discovered pikepdf which is more suitable for parsing the pdf inherit data structure. My knowledge about tagged pdfs is limited.
So, I have couple of questions:

1. How to parse the pikepdf structtreeroot representation?
2. Is it possible to dump it into a xml kind of structure, so it is easy for me to create a parser on top of it to extract Tables and other important tagged structures?
3. Can I can get Bounding boxes for these structures from the structTreeRoot itself? So, I can source link them back to pdf page; as I am used to with pymupdf. 

**Goal - to convert pdfs to simple text or json structure while utilizing the information from tagging.**
My intro to tagged pdfs was this - https://accessible-pdf.info/en/basics/general/overview-of-the-pdf-tags/#table-elements

**Comment by jbarlow83** [comment]
1. pikepdf can read structtreeroot, but it doesn't have the higher level code to interpret it and correlate it to document contents. For that you need to parse the content stream and look for BMC / BDC / EDC tags which mark the contents.

2. I suppose, but the effort to convert to XML then parse XML sounds like it would create more problems than it solves. The structtreeroot can be a cyclic graph since it can self-reference at certain points. (It's a really messy implementation within PDF and poorly implemented in some PDF software. I don't have many polite things to say about how it is designed, having spent a lot of time writing code to repair corrupted structtreeroots.)

3. You cannot get bounding boxes from structtreeroot. The main information it provides is the equivalent of markup. If you have access to Acrobat or Foxit, you can use the Accessibility and Tag tools, and see what they do; those tools manipulate structtreeroot and can also expose the structure in Preflight, which is essentially to visualizing what is going on.

To get bounding boxes you need something like pdfminer.six or pdfplumber (a wrapper around the former) which reads the fonts to establish their position. If the PDF does not include some of its fonts and they are not the basic 14, then font substitution applies and the PDF looks different to everyone.

**Comment by mllife** [comment]
@jbarlow83 , Can you suggest me some way to extract the tagged pdf information into a structured json or excel or something that is machine readable? I used foxit tmreader the view the tags but how to make it machine friendly format?

**Reply by jbarlow83** [reply]
You could use pikepdf's Object.to_json

**Reply by mllife** [reply]
@jbarlow83 , Can you suggest to some good resources or libraries/frameworks if I want to deep dive into the "StructTreeRoot" programatically?

**Reply by jbarlow83** [reply]
The PDF reference manual would be a good start.

**Reply by mllife** [reply]
this one? https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/pdfreference1.7old.pdf

**Reply by mllife** [reply]
@jbarlow83 , what tools you use to for programatic interaction with the pdf?
